React branch

evgdudareff opened the pull request:
Ветка с react

testUser346 committed:
Installed dependencies

testUser346 committed:
Added test less file

testUser346 committed:
Added .less support:

-installed less and less-loader
-replace scss with less extension in webpack.config (ejection command)

testUser346 committed:
Removed CRA configs and files. Moved project based on my own webpack.config and package.json

Also did common components (button, link, logo, navigation, layout). Start building header and navbar.

testUser346 committed:
Did Header block

testUser346 committed:
Did slider component

testUser346 committed:
Tuned localhost server to respond dev data (products info)

testUser346 committed:
Did content section

testUser346 committed:
Did features section

testUser346 committed:
Did footer

testUser346 committed:
Did content block with product-cards/popups

testUser346 committed:
Improved Input.js

testUser346 committed:
Did first form fields

testUser346 committed:
Did checkbox and improved Input block

testUser346 committed:
Did payment block

testUser346 committed:
Did select-block

testUser346 committed:
Improved controls configs

testUser346 committed:
Did textarea

testUser346 committed:
Did submit form button

testUser346 committed:
DidModal block and improved ProductCard/ProductCardPopup

testUser346 committed:
Did improvements

testUser346 committed:
Added redux and react-redux

testUser346 committed:
Did last improvements and ready for review

AndreyGladkov commented on src/components/Content/Content.js:
```diff
@@ -0,0 +1,34 @@
+export default class Content extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
```
В данном случае у тебя `productDataArr` - это props, записывая их в state, твои товары никогда не обновятся, потому как ты не описывал эту логику. 
В данном случае нет смысла в класс компоненте, достаточно функционального без использования state.

```
Object.values(props.productDataArr).map(
       (productData, index) => {
         return <ProductCard productData={productData} key={index} />;
       }
);
```

AndreyGladkov commented on src/components/Content/Content.js:
```diff
@@ -0,0 +1,34 @@
+  render() {
+    const products = Object.values(this.state.productDataArr).map(
+      (productData, index) => {
+        return <ProductCard productData={productData} key={index} />;
```
у тебя в данных есть `id`, в качестве ключа лучше использовать явные идентификаторы, а не полагаться на индекс массива.

AndreyGladkov commented on src/components/Content/Content.js:
```diff
@@ -0,0 +1,34 @@
+
+/* Компонент Content.
+Принимает: 
+- productDataArr      | array of object |    (массив объектов с данными товаров)
```
Для информации: если хочется описывать типы пропсов, есть удобный инструмент https://reactjs.org/docs/typechecking-with-proptypes.html

AndreyGladkov commented on src/components/Content/ProductCard/ProductCard.js:
```diff
@@ -0,0 +1,139 @@
+export default class ProductCard extends Component {
+  constructor(props) {
+    super(props);
+    this.onWindowResize = this.handleWindowResize.bind(this);
```
Для информации:
можно юзать синтаксический bind:
```
handleWindowResize = () => {//code}
```
Тогда не нужно будет привязывать контекст в конструкторе.

AndreyGladkov commented on src/components/Content/ProductCard/ProductCard.js:
```diff
@@ -0,0 +1,139 @@
+      productData: props.productData,
+      isCardOpened: false,
+      isMobile: false,
+      mobileSize: 768,
```
Данные которые не влияют на рендер лучше хранить в свойствах экземпляра класса.
Это помогает разграничить ответственность, данных от которых может произойти обновление жизненного цикла компонентов.

AndreyGladkov commented on src/components/Content/ProductCard/ProductCard.js:
```diff
@@ -0,0 +1,139 @@
+    this.setBodyOverflow = this.setBodyOverflow.bind(this);
+
+    this.state = {
+      productData: props.productData,
```
https://github.com/AndreyGladkov/hh-school-react-2020/pull/1/files#diff-6ca80e5354d6d73fd2a77164dc596e62R13

AndreyGladkov commented on src/components/Content/ProductCardPopup/ProductCardPopup.js:
```diff
@@ -0,0 +1,133 @@
+          <div className="product-card__button-submit">
+            <Button
+              title={"Заказать"}
+              modifierArr={submitButtonModifierArr}
```
Для информации: такие штуки лучше делать пропсами типо `kind={blue}`. который отвечал бы за стилистику по типу цветов, если нужно задать ширину 100% то это какой-нибудь `stretched={true}`

AndreyGladkov commented on src/components/Features/Feature/Feature.js:
```diff
@@ -0,0 +1,35 @@
+  const featureData = props.featureData;
+
+  return (
+    <GlobalContext.Consumer>
```
Для информации: в данном случае было удобнее юзать `useContext`

AndreyGladkov commented on src/components/Features/Feature/Feature.js:
```diff
@@ -0,0 +1,35 @@
+- featureData      | object |    (данные фичи)
+Использует глобальный контекст GlobalContext (путь до картинок)
+*/
+export default function Feature(props) {
```
Для информации: можно юзать диструктаризацию объекта и будет проще писать и читать:
```export default function Feature({featureData}) {//code}```

AndreyGladkov commented on src/components/Features/Features.js:
```diff
@@ -0,0 +1,42 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      features: [
```
нет смысла хранить это в state, в данном случае можно вынести в константу на уровне текущего компонента.

AndreyGladkov commented on src/components/Slider/Slide/Slide.js:
```diff
@@ -0,0 +1,52 @@
+- callToActionTitle | string |           (Заголовок для кнопки call to action)
+- activeSlideId     | string |           (Id/номер активного слайда)
+*/
+export default function Slide(props) {
```
Для информации: есть default props в купе с диструктаризацией могут значительно упростить написание и чтение:
```
{ src="",  srcset="", title="", callToActionTitle="Перейти"}
```

AndreyGladkov commented on src/components/common/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,75 @@
+    super(props);
+    this.onChangeHandlerSelf = this.onChangeHandlerSelf.bind(this);
+    this.state = {
+      id: Math.random(),
```
плохой идентификатор даже для html атрибута, в данном случае, лучше было бы завязаться на labelTitle, ну а если его нет, тогда и не нужен атрибут.

AndreyGladkov commented on src/components/common/Textarea/Textarea.js:
```diff
@@ -0,0 +1,55 @@
+        placeholder={this.state.placeholder}
+        name={this.state.name}
+        onChange={this.onChangeHandlerSelf}
+        value={this.state.value}
```
можно указать `defaultValue={this.props.value}` и тогда не нужно будет писать обработчик `this.onChangeHandlerSelf `

AndreyGladkov reviewed (changes requested):
Спасибо за решение. В целом все хорошо, за исключением нескольких моментов:
1. Не нужно хранить в state все что использует компонент, в state храним только данные, которые изменяются и которые влияют на рендер.
2. Не нужно класть в state пропсы компонента. Пропсы это внешние данные которые можно использовать render. Иначе нужно иметь логику обновления состояния после обновления props.
3. В качестве ключей использовать явные идентификаторы, если они есть. Стараться избегать ключей завязанных на индекс в массиве.

Я все отписал в пр, добавил комментарии для общих знаний. 
Главное поправить 3 пункта что я описал.

Молодец сам сделал свою сборку. Также что проявил интерес к контексту и заюзал его.

testUser346 committed:
Did corrections:

- redesigned state for compopnents (relocate no-state-properties)
- now props are not saved in state
- use organic keys where possible

AndreyGladkov commented on src/components/OrderForm/OrderForm.js:
```diff
@@ -154,19 +152,13 @@ export default class OrderForm extends Component {
-            modifierArr={
-              this.state.controls[deliveryGroup][controlName].modifierArr
-            }
+            name={this.controls[deliveryGroup][controlName].name}
```
Для информации: можно проще все эти пропсы прокидывать:
```
<Radio {...this.controls[deliveryGroup][controlName]} />
```

AndreyGladkov commented on src/components/common/Radio/Radio.js:
```diff
@@ -50,22 +43,23 @@ export default class Radio extends Component {
       className += " " + modifierArr.join(" ");
     }
 
+    const id = this.props.labelTitle + this.props.name + Math.random();
```
последнее значение в твоем случае лучше value

AndreyGladkov reviewed (approved):
Спасибо за исправления, ты молодец, все оперативно сделал и поправил.
